lib/puppet/resource_api/value_creator.rb

```diff
@@ -0,0 +1,87 @@
+module Puppet; module ResourceApi; end; end # predeclare the main module # rubocop:disable Style/Documentation,Style/ClassAndModuleChildren
+
+# This class is responsible for setting default and alias values for the
+# resource class.
+class Puppet::ResourceApi::ValueCreator
+  # This initialize takes arguments and sets up new value creator for given
+  # resource class which can be Puppet::ResourceApi::Parameter,
+  # Puppet::ResourceApi::ReadOnlyParameter, Puppet::ResourceApi::Property.
+  # @param resource_class the class of selected resource to be extended
+  # @param data_type the type instance
+  # @param definition the definition of the property
+  # @param options the ResourceAPI options hash containing setup information for
+  # selected parameter or property
+  def initialize(resource_class, data_type, param_or_property, options = {})
+    @resource_class = resource_class
+    @data_type = data_type
+    @param_or_property = param_or_property
+    @options = options
```

Comment:
all this state here seems like a lot of work for little benefit. Instead you could make this a module, and  `create_values` (and all other functions) `module_function`s and pass in the required data directly:

```ruby
def create_values(resource_class, data_type, param_or_property, options = {})
  # ...
```

With it being a module, you'd probably could even include it into the Property/Parameter classes directly and completely get rid of passing through the `resource_class`.

Reply:
Composition sounds like much better choice, thanks